### PR TITLE
Add a check for no CredentialsRequest manifest in directory 

### PR DESCRIPTION
**Pull request by thunderboltsid**
- Return an error to the user if the crederntials requests directory is empty
- Add unit tests for Nutanix create-shared-secrets subcommand

x-ref: https://issues.redhat.com/browse/CCO-108

### CODE REVIEW

File: pkg/cmd/provisioning/nutanix/create_shared_secrets.go [hunk 1]
```diff
@@ -166,7 +165,6 @@ func getCredentialsData(creds *NutanixCredentials) (*NutanixCredentials, error)
 		}
 	}
 
-	log.Printf("There are %d valid credentials", len(retCreds.Credentials))
```
Comment (thunderboltsid) [inline pkg/cmd/provisioning/nutanix/create_shared_secrets.go:169]: This log statement causes more confusion than it helps.

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    name: %s
+    namespace: %s`
+
+	basicAuthCredentialsTemplate = `
+credentials:
+- type: basic_auth
+  data:
+    prismCentral:
+      username: %s
+      password: %s`
```
Comment (akhil-rane) [inline pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:41]: Should it also include auth for prism elements as defined [here](https://github.com/openshift/cloud-credential-operator/blob/master/pkg/cmd/provisioning/nutanix/create_shared_secrets.go#L48-L56)? I understand it won't affect the tests here but just for consistency and correctness.

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+		{
+			name: "Non-existent source credentials file",
+			setup: func(t *testing.T) (credReqDir, targetDir, credentialsSourceFilepath string) {
+				credReqDir, err := ioutil.TempDir(os.TempDir(), testCredReqDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials requests")
+
+				targetDir, err = ioutil.TempDir(os.TempDir(), testTargetDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials requests")
+
+				credentialsSourceFilepath = "__does_not_exist__"
```
Comment (akhil-rane) [inline pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:103]: nit
```suggestion
credentialsSourceFilepath = "/does/not/exist"
```

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+				assert.Zero(t, countNonDirectoryFiles(files), "Should be no files in manifestsDir when no CredReqs to process")
+			},
+			expectedErr: "Failed to process files containing CredentialsRequests: open __does_not_exist__: no such file or directory",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			credReqDir, targetDir, credentialsSource := tt.setup(t)
+			defer os.RemoveAll(credReqDir)
+			defer os.RemoveAll(targetDir)
```
Comment (akhil-rane) [inline pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:138]: do we also need to add `defer os.RemoveAll(credentialsSource)`?

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+				require.NoError(t, err, "Failed to create temp directory for credentials requests")
+
+				targetDir, err = ioutil.TempDir(os.TempDir(), testTargetDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials requests")
+
+				credentialsDir, err := ioutil.TempDir(os.TempDir(), testCredentialsDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials")
+				credentialsSourceFilepath = testBasicAuthCredentials(t, "username", "password", credentialsDir)
+				return
+			},
```
Comment (akhil-rane) [inline pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:64]: `setup` function does not return anything, I am curious how this works. Is it based on variable names?

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go (hunk 2 above)
Comment (thunderboltsid) [reply pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:41]: Those are optional and currently, not in use. Let me see, I can add a test that tests the contents of the output too. Then we can have one with and one without.

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go (hunk 5 above)
Comment (thunderboltsid) [reply pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:64]: Indeed, these are named return parameters. The variables which will be returned are defined in the function signature.

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go (hunk 4 above)
Comment (thunderboltsid) [reply pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:138]: Good catch! Added it.

File: pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go (hunk 5 above)
Comment (akhil-rane) [reply pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:80]: thanks! learned something new :)

**Comment by jianping-shu** [comment]
@thunderboltsid @akhil-rane 
I updated the test case [OCP-48967](https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-48967) and added step 13-15 for this PR, PTAL!
One minor issue found in step 14 is I copied the following command during the test and there was a space at the end.
./ccoctl nutanix create-shared-secrets --credentials-requests-dir=./credentials_requests --output-dir=./nutanix_secrets  
Then ccoctl will include the space in the output dir name so it became 'nutanix_secrets'$'\302\240'.
Do you think we shall enhance this?

**Comment by thunderboltsid** [comment]
> @thunderboltsid @akhil-rane
> I updated the test case [OCP-48967](https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-48967) and added step 13-15 for this PR, PTAL!
> One minor issue found in step 14 is I copied the following command during the test and there was a space at the end.
> ./ccoctl nutanix create-shared-secrets --credentials-requests-dir=./credentials_requests --output-dir=./nutanix_secrets
> Then ccoctl will include the space in the output dir name so it became 'nutanix_secrets'$'\302\240'.
> Do you think we shall enhance this?

Jianping and I synced offline to reproduce this. I was unable to reproduce the issue highlighted with added space and the current working theory is the bash version. @jianping-shu will try reproducing with a more recent bash version. If we manage to reproduce this, we will address this in a separate PR.

**Comment by jianping-shu** [comment]
Talked w/ Sid in the slack, this issue has nothing to do with the PR, it will be further investigated and resolved in separate PR if need to enhance indeed.
/label qe-approved

**Pull request merged**